### PR TITLE
DEV: Allow setting a whitelist_group on LocalAuthenticator.

### PR DESCRIPTION
**Pull request by ssanderson**
Any user in the group is considered in the whitelist.

One issue that arises as a result of this PR is that it seems reasonable to have an empty whitelist of users if you're specifying a whitelist group, but that's currently not possible because the hub requires you to have an admin user and places them into the whitelist.

### CODE REVIEW

**Comment by ssanderson** [comment]
@minrk @YannBrrd I think this gets at most of https://github.com/jupyter/jupyterhub/issues/198

**Comment by minrk** [comment]
Empty whitelist is meant to be allowed, indicating that any user that successfully authenticates is permitted. Adding admin users to the whitelist should only happen if it is already non-empty. Can you verify that it's actually populating the whitelist?

File: jupyterhub/auth.py [hunk 1]
```diff
@@ -87,7 +94,28 @@ class LocalAuthenticator(Authenticator):
...
-    
+
+    whitelist_group = Unicode(
+        config=True,
+        help="Automatically whitelist anyone in this group.",
+    )
+
+    def check_whitelist(self, username):
+        return (
+            super().check_whitelist(username) or
```
Comment (minrk) [inline jupyterhub/auth.py]: In the case where the user whitelist is empty, `check_whitelist_group` will never be called.

For simplicity, I might require whitelisting by name or group to be mutually exclusive. What do you think?

**Comment by ssanderson** [comment]
@minrk sorry, I misunderstood what was happening in my initial test. I originally tried to test this by setting the whitelist to be empty, but, as you point out, that signifies "anyone is allowed".  I then tried to test it by setting the whitelist to be `{"root"}`, but then my user gets added as an admin and consequently added to the whitelist.

Ignoring the slightly confusion above, it seems like there should be a way of saying "don't allow any users by username: only allow them by group".  That's not currently possible because the empty set is interpreted as "allow anyone".  My inclination is to say that `None` should be used as "allow anyone", and  empty set should signify "no one is whitelisted".

File: jupyterhub/auth.py (hunk 1 above)
Comment (ssanderson) [inline jupyterhub/auth.py]: > In the case where the user whitelist is empty, check_whitelist_group will never be called.

This is the behavior I intended, though that doesn't necessarily mean it's correct.

I'm trying to think if there's ever a case where it's useful to be able to say "allow people in this group, plus these random other users"...I could imagine a world where you had something like an "admin" group on a server, and you wanted to allow access to all the admins, plus a few exceptions who shouldn't have admin permissions, but should still be able to access the hub.

**Comment by minrk** [comment]
@ssanderson if the group whitelist and name whitelist were mutually exclusive (I think they should be), then there shouldn't be a problem. You would *either* check by group *or* check by name.

**Comment by minrk** [comment]
Something like:

```python
if self.group_whitelist:
    return self.check_group_whitelist()
else:
    return self.check_whitelist()
```

File: jupyterhub/auth.py (hunk 1 above)
Comment (minrk) [inline jupyterhub/auth.py]: If the group whitelist is a set as well, I don't think there's anything you can do with names that you can't do with groups. Of course, a special-purpose 'jupyterhub' group would also cover it.

**Comment by ssanderson** [comment]
@minrk I updated this to make group and user whitelists mutually-exclusive.  Not sure what the deal is with the Python 3.3 failure on Travis.

File: jupyterhub/auth.py [hunk 2]
```diff
@@ -87,7 +94,34 @@ class LocalAuthenticator(Authenticator):
...
+        if self.group_whitelist:
+            return self.check_group_whitelist(username)
+        else:
+            return super().check_whitelist(username)
+
+    def check_group_whitelist(self, username):
+        if not self.group_whitelist:
+            return False
+        try:
+            group = getgrnam(self.group_whitelist)
```
Comment (minrk) [inline jupyterhub/auth.py]: If self.group_whitelist is a set now, doesn't this need to be a loop?

**Comment by minrk** [comment]
Don't worry about the test failure. There's a sporadic race condition that seems to cause it to fail once in a while. I haven't been able to track it down.

File: jupyterhub/auth.py (hunk 2 above)
Comment (ssanderson) [inline jupyterhub/auth.py]: Oops.  Yep. Will fix.

**Comment by minrk** [comment]
Thanks

**Pull request merged**